### ISSUE TITLE
GraphQL where filter possibilities

### ISSUE THREAD

**Issue by laura-ham** [creativesoftwarefdn/weaviate#630]
As @etiennedi noted out correctly, two issues arise regarding the `where` filter implementation of the GraphQL API.

### Or + nested NotEqual operators
First, queries like the following, where an `Or` operator is applied to two `NotEqual` filters are applied will always return all data. This is just basic logic. The question is whether we should provide this option to the user, because using this filter does not modify the returned results. On the other hand, applying a filter like this is not incorrect, so maybe returning the response enriched with a warning might be an idea to implement? This counts for Local as well as Network queries.

```graphql
{
  Local{
    Get(where:{
      operator: Or,
      operands: [{
        path: ["Things", "City", "name"],
        operator: NotEqual
        valueString: "Amsterdam"
      }, {
        path: ["Things", "City", "inCountry", "name"],
        operator: NotEqual,
        valueString: "Netherlands"
      }]
    }) {
      Things{
        City{
          name
        }
      }
    }
  }
}
```

### Network where filter paths
Currently, filters in the network are also placed on the `Get` level. In the `path` it should be specified to which Weaviate instance the filter operation should be applied to. However, it makes no sense to compare and combine nested filters of multiple Weaviate instances, because they come from different sources and are queried seperately in the next GraphQL query layer. @etiennedi would moving the `where` filter to the Weaviate instance/name level solve this problem? Then the filter will look (and work) exactly the same as the `Local` `where` filter. I currently see no advantage of having the filter on the `Get` level to compare property values of multiple Weaviates (except for Get(where) consistency), but I might be missing something.. I think one can just query and use filters per Weaviate instance to get desired results I guess.

This makes no sense:
``` graphql
{
  Network{
    Get(where:{
      operator: Or,
      operands: [{
        path: ["WeaviateB", "Things", "Airport", "place"],
        operator: Equal,
        valueString: "Amsterdam"
      }, {
        path: ["WeaviateC", "Things", "Human", "birthday"],
        operator: Equal,
        valueString: "01-01-01"
      }]
    }) {
      WeaviateB{
        Things{
          Airport{
            place
          }
        }
      }
      WeaviateC{
        Things{
          Human{
            birthday
          }
        }
      }
    }
  }
}
```

So proposed is:
```graphql
{
  Network{
    Get{
      WeaviateB(where: {
      	path: ["Things", "Airport", "place"],
        operator: Equal,
        valueString: "Amsterdam"
      }) {
        Things{
          Airport{
            place
          }
        }
      }
      WeaviateC(where:{
        path: ["WeaviateC", "Things", "Human", "birthday"],
        operator: Equal,
        valueString: "01-01-01"
    	}) {
        Things{
          Human{
            birthday
          }
        }
      }
    }
  }
}
```

**Comment by etiennedi** [comment]
@laura-ham 
> would moving the where filter to the Weaviate instance/name level solve this problem? Then the filter will look (and work) exactly the same as the Local where filter. I currently see no advantage of having the filter on the Get level to compare property values of multiple Weaviates (except for Get(where) consistency), but I might be missing something

Yes, this would be awesome. This would make quite some code that I wrote obsolete, which is always good. Less code => less complexity to maintain. The Network filtering would then really just be a very simple proxying, whereas right now because of having to remove the first element adds the overhead to parse an unknown structure, modify the path array, etc.

So I'm all for this change, both logically and implementation-wise.

There is a minor inconsistence in you example query though - I assume this is just a typo:
`Network.Get.WeviateB.where.path` does not start with the instance as the first element
`Network.Get.WeviateC.where.path` does start with `"WeaviateC"` which is redundant <-- I assume this wasn't intended.

Nevertheless, I'd still propose to first merge what we have now for Network.Get with the old filter structure and then tackle this as a separate concern slightly later to 
- make sure the branch doesn't deviate from`develop` to much.
- to allow me to make some refactoring on the graphql code first (to make the `Network` style more consistent with `Local`)
- to then (after refactoring!) split up the work between @JeroenStravers1 and me, so he can tackle the API change and I can tackle the resolver change. (If you're busy with something else or not available, I can also tackle both of course - just looking for most efficient workmode)

tl;dr: ✅ like the proposal, makes sense from a UX perspective, makes the code less complex than before, but want to tackle it only after merging the current implementation for `Network.Get`

**Comment by etiennedi** [comment]
@bobvanluijt @laura-ham @JeroenStravers1 

I've thought about this a bit more and actually I don't see any benefits of keeping the `where` filter at such a high level in any case (also not for `Local.Get`). What keeps us from moving the where filters all the way down to the individual class, like so?

```graphql
{
  Local {
    Get {
      Things {
        City (where: {
          operator: And,
          operands: [{
            path: ["population"],
            operator: GreaterThan
            valueInt: 1000000
          }, {
            path: ["inCountry", "Country", "name"],
            operator: Equal,
            valueString: "Netherlands"
          }]
        }){
          name
          InCountry {
            ... on Country {
              name
            }
          }
        }
      }
    }
  }
}
```

**What would we lose from this?**
- The ability to combine filters across different classes, however, I don't think they ever make sense. `All Cities with name not equal to Amsterdam OR all the cars with brand not equal to Mercedes` doesn't make sense from a logical perspective. They're just two seperate queries with two separate filters. 

**What would we gain from this?**
- The actual filtering is much closer to the Class I'm querying
- We don't have to specify redundant info anymore in the path, because the entire filter is already scoped to `["Thing", "City"]`, so we only need to keep the important stuff, such as `["population"]` or ` ["inCountry", "Country", "name"]`
- It would be consistent with the API design for grouping, where we [also apply the `groupBy` on a `Things.City` level](https://github.com/creativesoftwarefdn/weaviate/blob/develop/docs/en/use/graphql_grouping.md)

Am I missing anything?

**Comment by laura-ham** [comment]
@etiennedi I've thought about this for a bit more and I think your suggestion seems, for the reasons you mention, the most logical thing to do. I remember when I made the where filter, first I designed this in the way you propose (so the filter on the Class level). But I changed it to allow combinations of filters. Currently I however cannot come up with an example where you should need a combination in the filter of multiple classes, simply because it doensn't make sense from a logical perspective, as you point out.

Changing this would also solve #629. 

@bobvanluijt @JeroenStravers1 What do you think? Are we missing something important here or would you approve this change?

**Comment by JeroenStravers1** [comment]
I agree with both of you. As a user it would make more sense to me to apply filtering as close to the actual classes as possible, and I'm a big proponent of consistency in where elements are located. I can't think of any use case that hasn't already been discussed above.

**Comment by bobvanluijt** [comment]
First of all, before on of you will do it, the appropriate [link](http://m.memegen.com/bqpzgj.jpg) regarding my €0.02 on this topic.

I'm perfectly fine with this, _however_ I would like to understand the following implications first;
- @laura-ham Can you set this up in the docs and the prototype to keep the tests passing?
- @etiennedi do you have any insights on how much work this would be for you to make the actual changes?

**Comment by etiennedi** [comment]
@bobvanluijt 

>@etiennedi do you have any insights on how much work this would be for you to make the actual changes?

If I am not mistaken there are four things to change:
1. the graphql structure on `Local`
2. the graphql structure on `Network`
3. the graphql resolvers on `Local`
4. the graphql resolvers on `Network`

My assumption is 1 + 2 is just moving stuff around and making tests pass again, but @JeroenStravers1 would be the export on this, I'd say a few hours max.
For 3+4 we're already resolving at a later place anyway, so right now what we do is consume the filter where it appears and pass it through. All we have to do is move the passing through. I'd say a few hours max.

Overall should easily be achievable within a day unless I'm missing something (@JeroenStravers1).

**Comment by laura-ham** [comment]
@bobvanluijt

Same for me, its mainly moving some stuff, renaming a couple of GraphQL names and fields etc. Then the resolvers will take some more time indeed, but since the idea remains the same it won't take too long.

**Comment by JeroenStravers1** [comment]
@etiennedi is right, it's a matter of moving code around for me too

**Comment by etiennedi** [comment]
All parts are completed now including resolving for Local.Get and Network.Get. I'll submit a PR soon to close this issue.
